internal/uvm/scsi/mount.go

```diff
@@ -0,0 +1,146 @@
+package scsi
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sort"
+	"sync"
+
+	"github.com/Microsoft/hcsshim/internal/protocol/guestresource"
+)
+
+type mountManager struct {
+	m       sync.Mutex
+	mounter Mounter
+	// Tracks current mounts. Entries will be nil if the mount was unmounted, meaning the index is
+	// available for use.
+	mounts   []*mount
+	mountFmt string
+}
+
+func newMountManager(mounter Mounter, mountFmt string) *mountManager {
+	return &mountManager{
+		mounter:  mounter,
+		mountFmt: mountFmt,
+	}
+}
+
+type mount struct {
+	path       string
+	index      int
+	controller uint
+	lun        uint
+	config     *mountConfig
+	waitErr    error
+	waitCh     chan struct{}
+	refCount   uint
+}
+
+type mountConfig struct {
+	readOnly  bool
+	encrypted bool
+	verity    *guestresource.DeviceVerityInfo
+	options   []string
+}
+
+func (mm *mountManager) mount(ctx context.Context, controller, lun uint, c *mountConfig) (_ string, err error) {
+	// Normalize the mount config for comparison.
+	// Config equality relies on the options slices being compared element-wise. Sort the options
+	// slice first so that two slices with different ordering compare as equal. We assume that
+	// order will never matter for mount options.
+	sort.Strings(c.options)
+
+	mount, existed := mm.trackMount(controller, lun, c)
+	if existed {
+		<-mount.waitCh
```

Comment:
same here, we probably should wait on `ctx.Done` to allow for timeouts/cancels

Reply:
Changed.